File: MultipleViewTypeChat/app/src/main/java/com/xyz/multipleviewtypechat/adapter/ChatAdapter.java
package com.xyz.multipleviewtypechat.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.xyz.multipleviewtypechat.R;
import com.xyz.multipleviewtypechat.model.BaseModel;
import com.xyz.multipleviewtypechat.model.ReceiverChat;
import com.xyz.multipleviewtypechat.model.SenderChat;
import com.xyz.multipleviewtypechat.viewholders.ReceiverViewHolder;
import com.xyz.multipleviewtypechat.viewholders.SenderViewHolder;

import java.util.List;

public class ChatAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<BaseModel> baseModelList;

    public ChatAdapter(List<BaseModel> baseModelList) {
        this.baseModelList = baseModelList;
    }

    @NonNull
    @Override

    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = null;
        switch (viewType) {
            case 1:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.chat_sender_layout, parent, false);
                return new SenderViewHolder(view);
            case 2:
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.chat_receiver_layout, parent, false);
                return new ReceiverViewHolder(view);
        }
        return null;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {

        int viewType = baseModelList.get(position).getItemType();
        BaseModel model = baseModelList.get(position);
        switch (viewType) {
            case 1:
                if (holder instanceof SenderViewHolder) {
                    ((SenderViewHolder) holder).setSenderData((SenderChat) model);
                }
                break;
            case 2:
                if (holder instanceof ReceiverViewHolder) {
                    ((ReceiverViewHolder) holder).setData((ReceiverChat) model);
                }
                break;
        }
    }

    /**
     * This method returns the view type based on which we need to populate the view in
     * the onCreateViewHolder()
     *
     * @return if 1 then populate Sender layout , if 2 populate Receivers layout
     */
    @Override
    public int getItemViewType(int position) {
        return baseModelList.get(position).getItemType();
    }

    @Override
    public int getItemCount() {
        return baseModelList.size();
    }
}

File: FirebaseChat/app/src/main/java/com/xyz/firebasechat/adapter/ChatAdapter.kt
package com.xyz.firebasechat.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import com.xyz.firebasechat.R
import com.xyz.firebasechat.model.Chat
import com.xyz.firebasechat.viewholders.ChatReceiverViewHolder
import com.xyz.firebasechat.viewholders.ChatSenderViewHolder

class ChatAdapter(private var chatList: List<Chat>) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val ITEM_TYPE_SENT = 1
    private val ITEM_TYPE_RECEIVED = 2

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            ITEM_TYPE_SENT -> {
                val view =
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.chat_right_layout, parent, false)
                return ChatSenderViewHolder(view)
            }

            ITEM_TYPE_RECEIVED -> {
                val view =
                    LayoutInflater.from(parent.context)
                        .inflate(R.layout.chat_left_layout, parent, false)
                return ChatReceiverViewHolder(view)
            }
        }
        throw IllegalStateException("No matching views found")

    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val chat = chatList[position]
        if (holder is ChatSenderViewHolder) {
            holder.setData(chat)
        } else if (holder is ChatReceiverViewHolder) {
            holder.setData(chat)
        }
    }

    override fun getItemCount(): Int {
        return chatList.size
    }

    override fun getItemViewType(position: Int): Int {
        /*
        Check if the sender is sending the message or the receiver is sending and handle the view
        type accordingly
         */
        return if (chatList[position].sender == Firebase.auth.currentUser?.uid) {
            ITEM_TYPE_SENT
        } else {
            ITEM_TYPE_RECEIVED
        }
    }

    fun updateList(chatList: List<Chat>) {
        this.chatList = chatList
        notifyDataSetChanged()
    }
}